rclcpp/src/rclcpp/node_interfaces/node_base.cpp

```diff
@@ -213,6 +214,13 @@ NodeBase::create_callback_group(
 rclcpp::CallbackGroup::SharedPtr
 NodeBase::get_default_callback_group()
 {
+  // Create the default callback group, if needed.
+  if (nullptr == default_callback_group_) {
+    using rclcpp::CallbackGroupType;
+    default_callback_group_ =
+      NodeBase::create_callback_group(CallbackGroupType::MutuallyExclusive);
+  }
```

Comment:
This change makes it so `get_default_callback_group` is no longer logically `const`.  What's the particular reason for moving this out of the constructor and into here?

Reply:
The issue here is with the lambda that I used.  Since the lambda takes a weak pointer to node to later lock and get context, we need a valid weak pointer at the time that the callback group is created.

If the callback group is created in the constructor, smart pointers aren't valid yet (you can't get a shared or weak pointer in the constrcutor).

If you have a clever way around this, I would be willing to give it a shot.

Reply:
After a bit of thinking, I decided to remove the `has_valid_node` API and add some todo's about it.  As I discovered over here (https://github.com/ros2/geometry2/pull/595), while we shouldn't destroy a node before it's callback groups, there isn't any programmatic way of enforcing it right now.  Using the `has_valid_node` to throw an exception will cause users' programs to except where it previously didn't.

I think in the long term, we need to be better about managing Node/CallbackGroup lifetimes.  While the Node "owns" the default callback group (via shared pointer), it may share ownership with the person who constructed the Node (default callback group is a parameter).  Additionally, further added CallbackGroups are only held as weak_ptrs, so the node doesn't really "own" them.  This probably requires more consideration.

The TLDR for here is that we can drop the `has_valid_node` API (and corresponding checks in the entities collector) and just keep a weak copy of the context in the lambda to provide all that is necessary to create the guard conditions.